hack/generator/pkg/codegen/pipeline_create_arm_types.go

```diff
@@ -0,0 +1,464 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel/armconversion"
+	"github.com/pkg/errors"
+)
+
+// createArmTypesAndCleanKubernetesTypes walks the type graph and builds new types for communicating
+// with ARM, as well as removes ARM-only properties from the Kubernetes types.
+func createArmTypesAndCleanKubernetesTypes(idFactory astmodel.IdentifierFactory) PipelineStage {
+	return PipelineStage{
+		"Create ARM types and remove ARM-only properties from Kubernetes types",
+		func(ctx context.Context, definitions astmodel.Types) (astmodel.Types, error) {
+			// 1. Walk types and produce the new ARM types, as well as a mapping of Kubernetes Type -> Arm Type
+			// 2. Walk Kubernetes types, remove ARM-only properties and add conversion interface (use mapping
+			//    from step 1 to determine which ARM resource we need to convert to).
+			// 3. Merge results from 1 and 2 together. Add any type/definition from the originally provided
+			//    definitions that wasn't changed by the previous steps (enums primarily).
+
+			armTypes, kubeNameToArmDefs, err := createArmTypes(definitions)
+			if err != nil {
+				return nil, err
+			}
+
+			kubeTypes, err := modifyKubeTypes(definitions, kubeNameToArmDefs, idFactory)
+			if err != nil {
+				return nil, err
+			}
+
+			result := make(astmodel.Types)
+			result.Union(armTypes)
+			result.Union(kubeTypes)
+			for _, def := range definitions {
```

Comment:
Future @babbageclunk here - I had forgotten that Types.Add prevents overwriting! Which doesn't *necessarily* invalidate these comments but probably does - I'll leave them here anyway.

Reply:
I think with the other changes I've made (called now `TypesSetDisjointUnion`, now returns new map rather than modifying existing one), these issues are resolved. Will leave this comment unresolved though in case they're not.

Reply:
Yeah, I think that's good.